### ISSUE TITLE
python3 package should also have python39*.zip placed in the output with python39.dll so that embedded interpreters can find imports, etc.

### ISSUE THREAD

**Issue by MichaelGoulding** [microsoft/vcpkg#18719]
If you try to compile a project that embeds the python interpreter, Python needs to be able to find the python system libraries.

It looks in the following places:

In a debug build:

<path where process is located>\python39_d.zip;.\DLLs;.\lib;<path where process is located>

Can we get python3 to copy the appropriate python39.zip or python39_d.zip to the output along with the DLL?

**Comment by JackBoosY** [comment]
Confirmed.

**Comment by Osyotr** [comment]
Same with python2 I guess?

**Comment by JackBoosY** [comment]
I found that all the files in the compressed package are in Lib and are not compressed. Is this reasonable?

**Comment by Hoikas** [comment]
The project that embeds the interpreter needs to package the standard library in whatever format it expects. Not all embedding applications will use a zip file--some may use custom PEP 451 machinery--therefore it is inappropriate for the python3 port to enforce any decision in this regard.  The standard library can be found using FindPython's `Python_STDLIB` [result variable](https://cmake.org/cmake/help/latest/module/FindPython.html#result-variables). If you need a zip, you can create this as part of your build process, perhaps using CMake's [`file(ARCHIVE_CREATE`](https://cmake.org/cmake/help/latest/command/file.html#archive-create)

**Comment by MichaelGoulding** [comment]
Is there some way to make it easier to understand what to do?

For me, I added pybind11 to my project, but I would get what seemed to be obscure errors when I tried to run the resulting executable.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by JackBoosY** [comment]
@MichaelGoulding
After comparing the python3 installed by **vcpkg** with the python3 installed by the **installation package**, I found the following items:
1. The python script (.py) in _Lib_ is compressed into `python39.zip`.
2. Generated several extra binaries (venv*.exe and _freeze_importlib.exe).
3. Several `pyd` scripts prefixed with `_` are missing in the python3 root directory.
4. Dependent libraries are installed in the DLLs instead of in the root directory.
5. Added pip3 binary to Scripts.

So what I want to confirm is:
1. Whether the script installed in _Lib_ can be called directly without being compressed.
2. Whether the missing `pyd` script in item 3 is required.

**Comment by LilyWangLL** [comment]
I think this issue should be a port feature.

**Comment by Hoikas** [comment]
This issue is bogus and should be closed.

> 1. Whether the script installed in Lib can be called directly without being compressed.

python39.zip is only used by certain embedded interpreters. The default is for all the scripts in Lib to be uncompressed. Embedders are responsible for packaging their own Python stdlib; vcpkg should not interfere with this.

> 2. Whether the missing pyd script in item 3 is required.

All required pyd (DLL) files are included in the vcpkg distribution.

**Comment by LilyWangLL** [comment]
Ping @JackBoosY, could you help confirm this issue should be closed?

**Comment by MichaelGoulding** [comment]
If one links to the embedded python target, cmake should include it.
________________________________

**Comment by JackBoosY** [comment]
No we shouldn't close this issue since it's not fixed, the expected python tool should same with the pre-build one.

**Comment by Hoikas** [comment]
The vcpkg python tool is the same as the one installed by the installer on python.org

**Comment by MichaelGoulding** [comment]
> I think this issue should be a port feature.

I agree that a port feature might be appropriate for those using pybind to embed vs. those using pybind to create a native module.

**Comment by Osyotr** [comment]
> But it is different from the "Windows embeddable package" on the same download page on python.org and I suspect a lot of folks here are interested in the embedded version (I am)?

The official embeddable package does not provide any libraries to link to. You can use it only to *run* python scripts.
What you should really do is something like this (tested on linux, for windows you should do it at build step):
```
set(_py3_zip_temp_dir "${CMAKE_BINARY_DIR}/python_embeddable_zip_temp")
file(MAKE_DIRECTORY "${_py3_zip_temp_dir}")

install(
  CODE "set(_py3_zip_temp_dir \"${_py3_zip_temp_dir}\")"
  CODE "set(Python_STDLIB \"${Python_STDLIB}\")"
  CODE "set(Python_VERSION_MAJOR \"${Python_VERSION_MAJOR}\")"
  CODE "set(Python_VERSION_MINOR \"${Python_VERSION_MINOR}\")"
  CODE [[
    file(GLOB_RECURSE _py_stdlib_files RELATIVE "${Python_STDLIB}" "${Python_STDLIB}/*")
    list(FILTER _py_stdlib_files EXCLUDE REGEX "__pycache__")
    list(FILTER _py_stdlib_files EXCLUDE REGEX "^config*/.*$")
    list(FILTER _py_stdlib_files EXCLUDE REGEX "^idlelib/.*$")
    list(FILTER _py_stdlib_files EXCLUDE REGEX "^lib2to3/.*$")
    list(FILTER _py_stdlib_files EXCLUDE REGEX "^lib-dynload/.*$") # Copied separately
    list(FILTER _py_stdlib_files EXCLUDE REGEX "^lib-tk/.*$")
    list(FILTER _py_stdlib_files EXCLUDE REGEX "^tkinter/.*$")
    list(FILTER _py_stdlib_files EXCLUDE REGEX "(^|/)tests?/.*$")
    list(FILTER _py_stdlib_files EXCLUDE REGEX ".*\\.pyo$")
    execute_process(
      COMMAND "${CMAKE_COMMAND}" -E tar cf "${_py3_zip_temp_dir}/python${Python_VERSION_MAJOR}${Python_VERSION_MINOR}.zip" --format=zip ${_py_stdlib_files}
      WORKING_DIRECTORY "${Python_STDLIB}"
    )
    file(INSTALL DESTINATION "${CMAKE_INSTALL_PREFIX}/lib" FILES "${_py3_zip_temp_dir}/python${Python_VERSION_MAJOR}${Python_VERSION_MINOR}.zip")
  ]]
)

install(
  DIRECTORY "${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/$<$<CONFIG:DEBUG>:debug/>lib/python${Python_VERSION_MAJOR}.${Python_VERSION_MINOR}/lib-dynload"
  DESTINATION "${CMAKE_INSTALL_LIBDIR}/python${Python_VERSION_MAJOR}.${Python_VERSION_MINOR}"
  FILES_MATCHING
  PATTERN "*.debug" EXCLUDE
  PATTERN "*.so"
)
```

**Comment by GordonSmith** [comment]
@Osyotr - where did you get the ECLUDE REGEX items from?  I was building up a similar list via inspection between the python310.zip file and "Lib" contents?

Also I was planning on running `${Python3_EXECUTABLE} -m compileall -b . ` in the temp folder and excluding the `.py` files during the zip process?

...very helpful - thanks!...

**Comment by MichaelGoulding** [comment]
Replying to keep the issue open.

Thanks,
    Michael
________________________________
